crates/node-bindings/src/parcel/tracer.rs

```diff
@@ -0,0 +1,178 @@
+use napi::JsNumber;
+use napi::JsObject;
+use napi::JsString;
+use once_cell::sync::Lazy;
+use parking_lot::Mutex;
+use tracing_appender::non_blocking::WorkerGuard;
+
+/// Ensures there is only ever one tracer running, initialized the first time it is needed.
+/// When it's dropped certain resources related to tracing might flush to disk or be released.
+/// It should be kept alive for the duration of the program or the Parcel instance.
+static GLOBAL_TRACER: Lazy<Mutex<Option<Tracer>>> = Lazy::new(|| Default::default());
+
+#[derive(Debug)]
+pub enum TracerMode {
+  /// Disable the Tracer
+  Disabled,
+  /// Output the Tracer logs to Stdout
+  Stdout,
+  /// Output the Tracer logs to a file
+  File {
+    /// The directory where the log files will be written.
+    directory: String,
+    /// A prefix for the log file names.
+    prefix: String,
+    /// The maximum number of rotated files to keep.
+    max_files: u32,
+  },
+}
+
+/// Telling Napi how to convert from a JS type to the enum
+impl FromNapiValue for TracerMode {
+  unsafe fn from_napi_value(
+    env: napi::sys::napi_env,
+    napi_val: napi::sys::napi_value,
+  ) -> napi::Result<Self> {
+    let object = JsObject::from_napi_value(env, napi_val)?;
+    let mode = object.get_named_property::<JsString>("mode")?;
+    let v = match mode.into_utf8()?.as_str()? {
+      "disabled" => TracerMode::Disabled,
+      "stdout" => TracerMode::Stdout,
+      "file" => TracerMode::File {
+        directory: object
+          .get_named_property::<JsString>("directory")?
+          .into_utf8()?
+          .as_str()?
+          .to_string(),
+        prefix: object
+          .get_named_property::<JsString>("prefix")?
+          .into_utf8()?
+          .as_str()?
+          .to_string(),
+        max_files: object
+          .get_named_property::<JsNumber>("maxFiles")?
+          .get_uint32()?,
+      },
+      _ => return Err(napi::Error::from_reason("Invalid tracer mode")),
+    };
+    Ok(v)
+  }
+}
+
+/// Needs to exist for deserialization
+impl ToNapiValue for TracerMode {
+  unsafe fn to_napi_value(
+    _env: napi::sys::napi_env,
+    _val: Self,
+  ) -> napi::Result<napi::sys::napi_value> {
+    // we don't send this value back to JS
+    unimplemented!()
+  }
+}
+
+impl Default for TracerMode {
+  #[cfg(not(debug_assertions))]
+  fn default() -> Self {
+    Self::File {
+      directory: std::env::temp_dir()
+        .join("parcel_trace")
+        .to_string_lossy()
+        .to_string(),
+      prefix: "parcel-tracing".to_string(),
+      max_files: 4,
+    }
+  }
+
+  #[cfg(debug_assertions)]
+  fn default() -> Self {
+    Self::Stdout
+  }
+}
+
+/// A reference counted guard that holds the tracer
+/// open until all instances are dropped
+pub struct Tracer {
+  count: Arc<AtomicUsize>,
+  worker_guard: Arc<Option<WorkerGuard>>,
+}
+
+impl Tracer {
+  // The Tracer can only be initialized once per process instance,
+  //subsequent invocations will reuse an existing instance
```

Comment:
I don't understand why you need to handle this;

Reply:
Do we envision needing multiple different tracers running? For example during integration tests where they may need to output to files with different prefixes or something

Reply:
No I don't think so, but `tracing` itself won't allow more than one per process

Reply:
Updated